src/coreclr/src/System.Private.CoreLib/src/System/Threading/ProcessorIdCache.cs

```diff
@@ -0,0 +1,124 @@
+
+            return currentProcessorId;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static int GetCurrentProcessorId()
+        {
+            int currentProcessorIdCache = t_currentProcessorIdCache--;
+            if ((currentProcessorIdCache & ProcessorIdCacheCountDownMask) == 0)
+            {
+                return RefreshCurrentProcessorId();
+            }
+
+            return currentProcessorIdCache >> ProcessorIdCacheShift;
+        }
+
+        // If GetCurrentProcessorNumber takes any nontrivial time (compared to TLS access), return false.
+        // Check more than once - to make sure it was not because TLS was delayed by GC or a context switch.
+        internal static unsafe int ProcessorNumberSpeedCheck()
+        {
+            // NOTE: We do not check the frequency of the Stopwatch.
+            //       The frequency often does not match the actual timer refresh rate anyways.
+            //       If the resolution, precision or access time to the timer are inadequate for our measures here,
+            //       the test will fail anyways.
+
+            // default values pretend to be very long samples with default ratio
+            double minID = Stopwatch.Frequency * 50;  // 50 sec
+            double minTLS = Stopwatch.Frequency;      // 1 sec
+
+            // warm up the code paths.
+            UninlinedThreadStatic();
+            if (Thread.GetCurrentProcessorNumber() < 0)
+                return MaxIdRefreshRate;
+
+            long oneMicrosecond = Stopwatch.Frequency / 1000000;
+            for (int i = 0; i < 10; i++)
+            {
+                // we will measure at least 16 iterations and at least 1 microsecond
+                int iters = 16;
+                long t1 = 0;
+                while (t1 < oneMicrosecond)
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        Thread.GetCurrentProcessorNumber();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                }
+
+                minID = Math.Min(minID, (double)t1 / iters);
+
+                // we will measure at least 16 iterations and at least 1 microsecond
+                iters = 16;
+                t1 = 0;
+                while (t1 < oneMicrosecond)
+                {
+                    iters *= 2;
+                    t1 = Stopwatch.GetTimestamp();
+                    for (int j = 0; j < iters; j++)
+                    {
+                        UninlinedThreadStatic();
+                    }
+                    t1 = Stopwatch.GetTimestamp() - t1;
+                }
+
+                minTLS = Math.Min(minTLS, (double)t1 / iters);
+            }
+
+            s_processorIdRefreshRate = Math.Min((int)(minID / minTLS), MaxIdRefreshRate);
```

Comment:
I am not sure whether `minID / minTLS` is the best tradeoff for the refresh rate. I would make it `5 * minID / minTLS`.

Reply:
`5 * minID / minTLS`   will move the refresh rate on the typical nonfast machine (like x86) from 20-30 to 200-300. 
There must be a point when large refresh rate makes this API less useful than `ManagedThreadId`.  `ManagedThreadId` is relatively unique, so when threads exchange cores, they would have cache misses, but at least might not share (and ping-pong those cache lines). 
With `CoreID` and large refresh rate they could be actually sharing the data. And it would not be nice if another core swap happens within those 300 accesses.

I think that 300 is a bit too much and we should have as small a refresh rate as we think affordable. - I.E. if we have 30x slower ProcessorNumber, we should just amortize that over 30 accesses.

I did not try yet to validate/reject this theory. 
I need to think of an experiment to at least crudely estimate the effects of long refresh rates.

---
In reply to: [356995018](https://github.com/dotnet/runtime/pull/467#discussion_r356995018) [](ancestors = 356995018)

Reply:
Yeah, it is hard to get a conclusive answer for this. I can live with the 1:1 amortization ratio if you really believe that it is right. Or maybe take take average between our two opinions and make it 1:2 :-)

Reply:
I am trying to use this in a striped counter and so far it points to benefits of longer caching than 1:1.    Maybe  `5 * minID / minTLS ` is the right ratio... 
That is an extreme case though, very sensitive to the cost of access to a stripe, since uncontended increment is cheap.

Or maybe I have bugs somewhere :-)

---
In reply to: [357432015](https://github.com/dotnet/runtime/pull/467#discussion_r357432015) [](ancestors = 357432015)

Reply:
Here is the deal about refresh rate:
I tried using this API in a striped counter (https://github.com/VSadov/NonBlocking/tree/master/src/NonBlocking/Counter). The project has benchmarks so it was convenient to measure the impact.
Turned out that using this API vs. ManagedThreadId regressed the perf. Especially with low concurrency. At 2 threads it was roughly 2x slower. 
That was unexpected and gave ma a pause - I wanted to understand the result.

After some thinking I realized that there are too reasons why data structures use affinity.
1. To improve locality (avoid running on one core and using piece of data local to other core caches)
2. To reduce sharing (when two threads use the same piece of data)

Some scenarios, like a core-associated task queue in a thread pool, have large footprint and sensitive to both parts. It is desirable to use the data from the "right" core.
In scenarios where the state is small (like a counter), it is mostly about sharing. Otherwise state is small and occasionally moving counter to a different core via cache miss is not a big deal.

For scenarios that care only about sharing ManagedThreadId works just fine. In fact even that maybe an overkill. Anything thread-specific (like an address of a local) would work. CoreID does not pay for itself in this scenario unless definitely cheaper than ManagedThreadId.

To test the theory I added a 256 elements int array to each counter stripe and made it to increment the elements of the array when incrementing a counter. With such addition, locality started to matter more and CoreID has become competitive.
CoreID was a mild-to-definite win on a machine with RDPID, depending on concurrency. On a machine with RDTSCP-based CoreID, the extra cost was still keeping it from being a win, most notably at low thread counts.

So, I enabled caching. At 5x amortization ratio RDTSCP machine has refresh rate at roughly 10x and CoreID becomes a parity or better than `ManagedThreadId` depending on thread count. Longer caching does not improve results by much. 

Overall, the benefits of caching depend on the data structure, how it is used and on concurrency level. I.E. if an array pool user only rents array "just in case" but does not actually use it, and concurrency level is low, longer refresh would be beneficial since that could lower the cost. If array is actually used, then there is benefit from higher precision of the API and shorter refresh.

We can not know the ideal refresh rate and deriving it from some kind of dynamic feedback is unlikely to pay for the costs. 
I think `5x` amortization rate could be good enough for most purposes. 

Also, RDPID is still detected as "fast" and bypasses cache, since caching is not an improvement on RDPID machine.